### DISCUSSION TITLE
Reactive Quarkus gRPC Interceptor / How to interact with non-blocking call in ServerInterceptor

### DISCUSSION THREAD

**Discussion by paiyar** [quarkusio/quarkus#34322] [Q&A] [answered]
Given a reactive non-blocking auth service like so:

```JAVA
public interface AuthService {
    Uni<Claims> verify(String token);
}
```

How does one implement a gRPC interceptor that invokes a reactive / non-blocking methods e.g.:

```JAVA
@GrpcService
@RegisterInterceptor(AuthInterceptor.class)
public class GrpcService implements Api {
    public Uni<Response> request(Response response) {}
}

@ApplicationScoped
public class AuthInterceptor implements ServerInterceptor {

    @Inject
    AuthService authService;

    @Override
    public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
            ServerCall<ReqT, RespT> serverCall,
            Metadata metadata,
            ServerCallHandler<ReqT, RespT> serverCallHandler
    ) {
        // Obviously we cannot block here or Vert.x will pitch a fit...
        authService.verify(metadata.get(AUTHORIZATION_HEADER_KEY)).await().indefinitely();
        return serverCallHandler.startCall(serverCall, metadata);
    }
}
```

[a comment by an automated account is not shown]

**Comment by paiyar** [comment]
This is related to a similar question about calling  non-blocking `Uni` methods from an HTTP client interceptor: https://github.com/quarkusio/quarkus/discussions/28637

@cescoffier , @manofthepeace Sorry to bother, could you please advise?

**Reply by alex-martel** [reply, marked as answer]
@paiyar I have answered the followup question and added a sample app in the other discussion you linked. Generally I would suggest not using an interceptor for that purpose, but the app I have linked shows does work. Basically the thing is that the subscription must not happen in the interceptor. hope that helps.

**Reply by paiyar** [reply]
Thank you @manofthepeace, appreciate the example, I agree with your concerns about type safety and think I'll avoid the interceptor approach (despite how much cleaner it may have been in theory to abstractly enforce JWT claim based authZ in an AuthInterceptor)
